pkg/driver/driver_azure.go

```diff
@@ -167,14 +171,35 @@ func (d *AzureDriver) getVMParameters(vmName string, networkInterfaceReferenceID
 		VMParameters.Identity = &compute.VirtualMachineIdentity{
 			Type: compute.ResourceIdentityTypeUserAssigned,
 			UserAssignedIdentities: map[string]*compute.VirtualMachineIdentityUserAssignedIdentitiesValue{
-				*d.AzureMachineClass.Spec.Properties.IdentityID: &compute.VirtualMachineIdentityUserAssignedIdentitiesValue{},
+				*d.AzureMachineClass.Spec.Properties.IdentityID: {},
 			},
 		}
 	}
 
 	return VMParameters
 }
 
+func (d *AzureDriver) generateDataDisks(vmName string, azureDataDisks []v1alpha1.AzureDataDisk) []compute.DataDisk {
+	var dataDisks []compute.DataDisk
+	for _, azureDataDisk := range azureDataDisks {
+		dataDiskName := dependencyNameFromVMNameAndDependency(getAzureDataDiskName(azureDataDisk), vmName, dataDiskSuffix)
+		dataDiskLun := azureDataDisk.Lun
+		dataDiskSize := azureDataDisk.DiskSizeGB
+		dataDisk := compute.DataDisk{
+			Lun:     &dataDiskLun,
```

Comment:
Can we simply enumerate the `lun` starting with `0`? This will be only called when a new machine is created. In all other cases we would need to determine a free `lun` first because k8s could already attach a data disk which allocate the `lun`.

Reply:
well, it depends on the use case. If MCM is used purely for declarative VM management (outside of k8s), it might  be useful to configure luns explicitly. For example, the recommended way for [identifying devices within the VM is by lun](https://docs.microsoft.com/en-us/azure/virtual-machines/troubleshooting/troubleshoot-device-names-problems#identify-disk-luns). This is also consistent with the device-naming behavior in the AWS machine controller.

Reply:
Hmm ok, you mean machines which are managed as Kubernetes resources but not intended to be k8s nodes. Makes sense for me. @prashanth26, @hardikdr WDYT?

Could we then make the `lun` optional? In case no `lun` is set then enumerate and starting with 0? WDYT?

Reply:
Well, this input is used both for disk creation, deletion and on-error cleanup flows. so, lun ids will need to be generated in several scenarios. If it was possible to default this value before storage, handling unset luns would be much cleaner. To avoid over complication, I suggest that luns would either be exclusively set or unset.  WDYT ?

Reply:
I think I am fine exclusively setting/unsetting it for the scope of this PR.
Though, we might want to in the future allow 
1. optional lun, if not specified start from 0, as pointed by @dkistner .
2. And check on the free lun before allocating, in case we decide to move towards managing machines not necessarily being k8s nodes.

Reply:
> How are the luns managed by the PVC/PV controller in k8s ?

The azure k8s driver is fetching the machine spec from Azure to check for free `luns` before attaching a volume. Therefore it shouldn't be possible that k8s allocate the same `lun` for data disk which is attached to machine by mcm.

My suggestion was to make the `lun` in the spec optional. If set ok then use that one on the machine (if invalid machine creation will of course fail). If not set start to enumerate the lun beginning with 0.

Reply:
>My suggestion was to make the lun in the spec optional. If set ok then use that one on the machine (if invalid machine creation will of course fail). If not set start to enumerate the lun beginning with 0.

Understood thanks. Having optional lun would be indeed nice. 
Though, the current state doesn't seem to be hurting, leaving the decision to @guydaichs to change the behavior.

Reply:
Considering that the Azure Disk Deletion API is based on resource names https://docs.microsoft.com/en-us/rest/api/compute/disks/delete, it's important to establish unique and stable names for data disks. One option is to use the already-unique lun attribute in the naming convention, to achieve this uniqueness. If there was an option to mutate the machine class and set a value for unspecified luns once, that would fine. But we don't have this ability in MCM yet, and I'm concerned that generating this unique identifier in several flows can lead to inconsistency. For the scope of this PR, I prefer to keep the MCM simple and  require this value to be set.

Reply:
Overall, sounds good to me. MCM could potentially lose the allocated-lun and storing it at MachineClass dynamically, could actually bring complexity, and probably not in the scope of this PR. 
@dkistner closing the comment, please feel free to re-open if you think otherwise :)

Reply:
Hmm I try to understand the concerns. The data disks are exclusively bound to the machine and can't be reused by other machines. That mean the disks are also bound to the lifecycle of the machine. 

On creation of the machine the data disks will be also created and at this point in time all luns free on the machine, so we could start with 0 lun and enumerate for other data disks.
During lifetime k8s will attach new disks, but k8s (or someone else) will also check for free luns before attaching the disk which is anyways the safest way to attach new disks.
If the MachineClass spec will change (e.g. adding a new data disks) then we would need to roll the machines and new data disks, which contain the name of the new machine in their name, will be created, right?
On deletion of the machine the attached data disks will be deleted together with machine.

Name generation for the data disk shouldn't be also a big deals as we already have a `name` for each data disk. In combination with machine name it should be unique. We only need to validate that the data disk names are unique.

So what's the concern? Losing a disk (e.g. when something unexpected like mcm crashes happen during the deletion process) which would lead to an orphan disk? That could be handled by the safety controller and by adding some tags to the data disk which contain the machine name. So if the machine is gone and belonging data disks are still exists then it could find and delete them.

Don't get me wrong. The proposed approach is ok for the moment, but from my point of view `lun` is an internal detail of the machine and for connivence we should not force the user to specify it. Users should simply need to say I want a data disk, with this name, size and storage type (+ maybe I need also caching). But I understand this is more effort to archive.

Reply:
I overall agree with the suggestion to keep the lun optional. But I also feel, having luns as part of the disk-names makes the uniqueness stronger. As we don't want to add the complexity of updating the machineclass with generated-luns, let's keep it the current way for now. We can always have an another look at it in the future if we see a need.